src/Accelerators/NNPA/Transform/ZLow/ZLowRewrite.cpp

```diff
@@ -100,6 +112,330 @@ class StickViewUnstickRemovalPattern : public OpRewritePattern<ZLowStickOp> {
+        else
+          opToMove->moveAfter(storeAllocOp);
+        justMovedOp = opToMove;
+      }
+      // Move AllocOp.
+      if (justMovedOp)
+        stickAllocOp->moveAfter(justMovedOp);
+      else
+        stickAllocOp->moveAfter(storeAllocOp);
+
+      // Replace storeOp.
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointAfter(storeOp);
+      // This DummyOp is used to make the intermediate generated code valid. It
+      // will be removed automatically via canonicalization.
+      Value dummyConverter =
+          rewriter.create<ZLowDummyOp>(loc, stickifiedElementType, storeValue);
+      // Clone storeOp with new Memref, Value, and Indices.
+      IRMapping operandMap;
+      operandMap.map(storeOp.getMemref(), stickMemref);
+      operandMap.map(storeOp.getValue(), dummyConverter);
+      // Permute indices in case of NCHW layout.
+      if (isNCHWLayout) {
+        // for zlow.stick: input is NCHW, output is NHWC.
+        ValueRange NCHWIndices = storeOp.getIndices();
+        SmallVector<Value, 4> NHWCIndices;
+        NHWCIndices.emplace_back(NCHWIndices[0]); // N
+        NHWCIndices.emplace_back(NCHWIndices[2]); // H
+        NHWCIndices.emplace_back(NCHWIndices[3]); // W
+        NHWCIndices.emplace_back(NCHWIndices[1]); // C
+        operandMap.map(NCHWIndices, NHWCIndices);
+      }
+      rewriter.clone(*storeOp.getOperation(), operandMap);
+      rewriter.eraseOp(storeOp);
+    }
+
+    // Remove ZLowUnstickOp.
+    rewriter.eraseOp(unstickOp);
+    // Copy ZLowStickOp to the removableStickOps. We cannot remove it now
+    // because there are potentially other AffineStoreOps connecting to it.
+    for (ZLowStickOp stickOp : stickOps)
+      removableStickOps.insert(stickOp);
+    return success();
+  }
+
+private:
+  llvm::SmallDenseSet<ZLowStickOp, 4> &removableStickOps;
+
+  // Collect affine.load operations that connect to zlow.unstick.
+  bool matchAndCollectAffineLoad(ZLowUnstickOp unstickOp, Value loadMemref,
+      SmallVectorImpl<AffineLoadOp> &loadOps) const {
+    for (Operation *user : loadMemref.getUsers()) {
+      if (user == unstickOp.getOperation())
+        continue;
+      if (auto loadOp = llvm::dyn_cast<AffineLoadOp>(user))
+        loadOps.emplace_back(loadOp);
+      else
+        return false;
+    }
+
+    return (loadOps.size() != 0);
+  }
+
+  // Collect affine.store operations that connect to affine.load.
+  bool matchAndCollectAffineStore(const SmallVectorImpl<AffineLoadOp> &loadOps,
+      SmallVectorImpl<AffineStoreOp> &storeOps) const {
+    for (AffineLoadOp loadOp : loadOps) {
+      Value loadValue = loadOp.getValue();
+      for (Operation *user : loadValue.getUsers()) {
+        if (user == loadOp.getOperation())
+          continue;
+        if (auto storeOp = llvm::dyn_cast<AffineStoreOp>(user)) {
+          // Store's input must be defined by a memref.alloc.
+          Value storeMemref = storeOp.getMemref();
```

Comment:
This constraint will not exist if this transformation is done before bufferization.   Right?

Reply:
Yes, bufferization would be done in advance.